pysteps/verification/sal.py

```diff
@@ -0,0 +1,357 @@
+# -- coding: utf-8 --
+"""
+pysteps.verification.sal
+==================================
+
+The Spatial-Amplitude-Location (SAL) score.
+
+.. autosummary::
+    :toctree: ../generated/
+
+    sal
+    sal_structure
+    sal_amplitude
+    sal_location
+"""
+from math import sqrt, hypot
+
+import numpy as np
+from scipy.ndimage.measurements import center_of_mass
+
+from pysteps.exceptions import MissingOptionalDependency
+from pysteps.feature import tstorm as tstorm_detect
+
+try:
+    import pandas as pd
+
+    pandas_imported = True
+except ImportError:
+    pandas_imported = False
+
+try:
+    from skimage.measure import regionprops_table
+
+    skimage_imported = True
+except ImportError:
+    skimage_imported = False
+
+
+def sal(
+    prediction,
+    observation,
+    tstorm_kwargs=None,
+):
+    """
+    Compute the Structure-Amplitude-Location (SAL) spatial verification metric.
+
+    Parameters
+    ----------
+    prediction: array-like
+        Array of shape (m,n) with prediction data.
+    observation: array-like
+        Array of shape (m,n)  with observation data.
+    tstorm_kwargs: dict, optional
+        Optional dictionary containing keyword arguments for the tstorm feature
+        detection algorithm.
+        See the documentation of :py:mod:`pysteps.feature.tstorm`.
+
+    Returns
+    -------
+    sal: tuple of floats
+        A tuple of floats containing the structure, amplitude, location components of SAL.
+
+    References
+    ----------
+    :cite: Wernli, H., Hofmann, C., & Zimmer, M. (2009).
+    :cite: Feldmann, M., Germann, U., Gabella, M., & Berne, A. (2021).
+
+    Notes
+    -----
+    This implementation uses the thunderstorm detection algorithm by Feldmann et al (2021)
+    for the identification of precipitation objects within the considered domain.
+
+    This approach uses multi-threshold algorithm to detect objects, instead of having a
```

Comment:
@EsmailGhaemi do you think it would be difficult to implement the original approach too? Would be nice to actually give the option to use both, as tstorm requires a bit more attention in specifying the parameters ...

Reply:
See c6022c2.

@feldmann-m you think it makes sense to set the parameters of tstorm.detection as

```
        tstorm_kwargs = {
            "minsize":  5,
            "minmax": 0,
            "mindis": 5,
            "maxref": np.Inf,
            "minref": 1/15 * np.nanmax(precip),
        }
```

to obtain a simpler object detection based solely on a detection threshold (1 / 15 * precip_max)?

Reply:
> @EsmailGhaemi do you think it would be difficult to implement the original approach too? Would be nice to actually give the option to use both, as tstorm requires a bit more attention in specifying the parameters ...

Hi and thanks a lot for making it way better :) It is a good idea if we have the original approach using f value as well. I think it would be better to have a separate function for that. For instance, sal_single_threshold. If you agree, I'll add a function for object detection using f value here.

Reply:
Hi @EsmailGhaemi, I tried to implement the single threshold approach in commit c6022c2. Please have a look at the code and see if it makes sense to you.

Reply:
Yes, fantastic! 
However, Wernli et al., (2009) suggested to use the 95th quantile (larger than 0.1 mm), instead of maximum value. If you agree, I'll change it to 95th quantile ((1/15) * precip_95quantile).

Reply:
Of course, please do